meinberlin/apps/projects/tasks.py

```diff
@@ -0,0 +1,107 @@
+        .first()
+    )
+    if not phase:
+        return logger.warning("At least one phase with dates is required")
+    return phase.start_date.strftime("%Y-%m-%d, %H:%M:%S")
+
+
+def get_next_project_ends() -> str:
+    """
+    Helper function to query the db and
+    retreive the erliest phase that will end.
+    """
+    phase = (
+        Phase.objects.filter(end_date__isnull=False, end_date__gt=timezone.now())
+        .order_by("end_date")
+        .first()
+    )
+    if not phase:
+        return logger.warning("At least one phase with dates is required")
+    return phase.end_date.strftime("%Y-%m-%d, %H:%M:%S")
+
+
+@shared_task(name="schedule_reset_cache_for_projects_becoming_active")
+def schedule_reset_cache_for_projects_becoming_active():
+    """The task is set via celery beat every 10 minutes in
+    settings/production.txt.
+
+    Returns:
+        A string message if the next project starts
+        date will be removed from the cache.
+    """
+
+    # check if redis has a next_project_starts key
+    next_project_starts = cache.get("next_project_starts")
+
+    if not next_project_starts:
+        # set the redis key: value
+        cache.set("next_project_starts", get_next_project_starts())
+
+    # compare now with next start date
+    project_starts = datetime.strptime(next_project_starts, "%Y-%m-%d, %H:%M:%S")
+    remain_time = project_starts - timezone.now()
+    if remain_time <= timedelta(minutes=10):
+        # schedule cache clear for the remaining time btwn now and next start
+        reset_cache_for_projects.apply_async([True, False], countdown=remain_time)
+        msg = logger.info(
+            f"""next project starts date {next_project_starts}
+            will be removed from cache in {remain_time} min"""
+        )
+        return msg
+
+
+@shared_task(name="schedule_reset_cache_for_projects_becoming_past")
+def schedule_reset_cache_for_projects_becoming_past():
+    """The task is set via celery beat every 10 minutes in
+    settings/production.txt.
+
+    Returns:
+        A string message if the next project ends
+        date will be removed from the cache.
+    """
+
+    # check if redis has a next_project_ends key
+    next_project_ends = cache.get("next_project_ends")
+    if not next_project_ends:
+        cache.set("next_project_ends", get_next_project_ends())
+
+    # compare now with next end date
+    project_ends = datetime.strptime(next_project_ends, "%Y-%m-%d, %H:%M:%S")
+    remain_time = timezone.now() - project_ends
+    if remain_time <= timedelta(minutes=10):
+        # schedule cache clear for the remaining time btwn now and next end
+        reset_cache_for_projects.apply_async([False, True], countdown=remain_time)
+        msg = logger.info(
+            f"""next project ends date {next_project_ends}
+            will be removed from cache in {remain_time} min"""
+        )
+        return msg
+
+
+@shared_task
+def reset_cache_for_projects(starts, ends):
+    if starts:
```

Comment:
Do we have to call schedule_reset_cache_for_projects_* here again or run this for all projects starting within 10 minutes ? Considering the following case:
1. Project A starts at 12:01, Project B starts at 12:03
2. The schedule task runs at 12:00 and now schedules a reset_cache_for_projects for 12:01
3. At 12:01 the cache is reset and now includes Project A
4. At 12:03 Project B won't be visible yet because schedule_reset only runs at 12:10
5. At 12:10 schedule_reset runs again, if there's no future project starting the cache won't be updated and Project B still won't be visible, if there is a future Project C, Project B will only become visible once Project C starts

This is more complicated then I hoped :/ Not sure how to make this bulletproof. Maybe we can have a look together?

Reply:
Quite a niche case, but valid point nonetheless. I will check for a list of future projects that are to be published in the next 10 mins with the function `get_next_project_starts()` and apply the cache reset for that list of projects.